python/ql/src/experimental/dataflow/TypeTracker.qll

```diff
@@ -0,0 +1,282 @@
+/** Step Summaries and Type Tracking */
+
+import python
+import internal.DataFlowPublic
+import internal.DataFlowPrivate
+
+/** Any string that may appear as the name of an attribute or access path. */
+class AttributeName extends string {
+  AttributeName() { this = any(Attribute a).getName() }
+}
+
+/** Either an attribute name, or the empty string (representing no attribute). */
+class OptionalAttributeName extends string {
```

Comment:
It feels like there should be QL class inheritance, e.g. `AttributeName extends OptionalAttributeName`. Could make `OptionalAttributeName` abstract and have `class MissingAttributeName extends OptionalAttributeName`.

Reply:
The present situation mirrors what the JS libraries do, however as we're already diverging somewhat from that bit of code, it's probably worthwhile to rethink this. In this case, I'm tempted to turn it into an IPA type instead. I imagine the current construction was made before these became available (and the use of `""` as an indicator of "no attribute" already bugged me a bit).

Reply:
Hmm... Having tried this, I'm not sure I think it's an improvement. You end up with a lot of packing and unpacking of the extra IPA type, and I think this overall makes the code harder to read.